### ISSUE TITLE
ANDROID_NDK_REVISION regex is wrong

### ISSUE THREAD

**Issue by DanAlbert** [android-ndk/ndk#596]
davidben pointed this out to me:

```bash
$ cat CMakeLists.txt 
message("ANDROID_NDK_REVISION is ${ANDROID_NDK_REVISION}")

$ cmake -DCMAKE_TOOLCHAIN_FILE=~/android-ndk-r16b/build/cmake/android.toolchain.cmake
-- Check for working C compiler: /usr/local/google/home/davidben/android-ndk-r16b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang
-- Check for working C compiler: /usr/local/google/home/davidben/android-ndk-r16b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/google/home/davidben/android-ndk-r16b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++
-- Check for working CXX compiler: /usr/local/google/home/davidben/android-ndk-r16b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
ANDROID_NDK_REVISION is 161.4479499
```

That should be 16.1.4479499.

**Comment by alalek** [comment]
This should fix this problem (add dot '.' after `\\1`)
```.diff
 # Android NDK revision
 file(READ "${ANDROID_NDK}/source.properties" ANDROID_NDK_SOURCE_PROPERTIES)
 set(ANDROID_NDK_SOURCE_PROPERTIES_REGEX
   "^Pkg\\.Desc = Android NDK\nPkg\\.Revision = ([0-9]+)\\.")
 if(NOT ANDROID_NDK_SOURCE_PROPERTIES MATCHES 
 "${ANDROID_NDK_SOURCE_PROPERTIES_REGEX}")
   message(SEND_ERROR "Failed to parse Android NDK revision: 
 ${ANDROID_NDK}/source.properties.\n${ANDROID_NDK_SOURCE_PROPERTIES}")
 endif()
-string(REGEX REPLACE "${ANDROID_NDK_SOURCE_PROPERTIES_REGEX}" "\\1"
+string(REGEX REPLACE "${ANDROID_NDK_SOURCE_PROPERTIES_REGEX}" "\\1."
  ANDROID_NDK_REVISION "${ANDROID_NDK_SOURCE_PROPERTIES}")
```

**Comment by alalek** [comment]
BTW, There is another "android.toolchain.cmake" from Android Studio's custom CMake package (cmake/3.6.4111459/android.toolchain.cmake).

It handles version better:
```
file(READ "${ANDROID_NDK}/source.properties" ANDROID_NDK_SOURCE_PROPERTIES)
set(ANDROID_NDK_SOURCE_PROPERTIES_REGEX
	"^Pkg\\.Desc = Android NDK\nPkg\\.Revision = ([0-9]+\\.[0-9]+\\.[0-9]+(-beta[0-9]+)?)\n$")
if(NOT ANDROID_NDK_SOURCE_PROPERTIES MATCHES "${ANDROID_NDK_SOURCE_PROPERTIES_REGEX}")
	message(FATAL_ERROR "Failed to parse Android NDK revision: ${ANDROID_NDK}/source.properties.")
endif()
string(REGEX REPLACE "${ANDROID_NDK_SOURCE_PROPERTIES_REGEX}" "\\1"
	ANDROID_NDK_PACKAGE_REVISION "${ANDROID_NDK_SOURCE_PROPERTIES}")
if(NOT ANDROID_NDK_PACKAGE_REVISION MATCHES "^${ANDROID_NDK_REVISION}\\.")
	message(FATAL_ERROR "Invalid Android NDK revision (should be ${ANDROID_NDK_REVISION}): ${ANDROID_NDK_PACKAGE_REVISION}.")
endif()
```

But Android Studio build tools doesn't pick this toolchain from bundled CMake.

**Comment by DanAlbert** [comment]
Yeah, the one in Studio is actually older. I broke this while trying to fix a different issue (-canary wasn't handled). Will have a fix for r17.

**Comment by DanAlbert** [comment]
https://android-review.googlesource.com/#/c/platform/ndk/+/585851

I've changed the semantics here slightly.
    
* ANDROID_NDK_REVISION is now the string we get from the file. This potentially includes "-beta1" or "-beta2", so may not be suitable for direct version comparisons.
    
* Exposed ANDROID_NDK_MAJOR, ANDROID_NDK_MINOR, ANDROID_NDK_BUILD, and ANDROID_NDK_BETA for numeric values for the individual components. Whether or not this is a canary build is not exposed because that information is not present in the source.properties (iirc we tried adding it and it made Studio unable to parse the file).

If anyone objects to those changes, let me know.
